lib/Dialect/LLHD/Simulator/Engine.cpp

```diff
@@ -105,40 +116,44 @@ int Engine::simulate(int n, uint64_t maxTime) {
     if (traceMode >= 0)
       trace.flush();
 
-    // Apply the signal changes and dump the signals that actually changed
-    // value.
-    for (auto change : pop.changes) {
-      // Get a buffer to apply the changes on.
-      Signal *curr = &(state->signals[change.first]);
+    // Process signal changes.
+    size_t i = 0, e = pop.changesSize;
+    while (i < e) {
+      const auto sigIndex = pop.changes[i].first;
+      const auto &curr = state->signals[sigIndex];
       APInt buff(
-          curr->size * 8,
-          ArrayRef<uint64_t>(reinterpret_cast<uint64_t *>(curr->value.get()),
-                             curr->size));
-
-      // Apply all the changes to the buffer, in order of execution.
-      for (auto drive : change.second) {
-        if (drive.second.getBitWidth() < buff.getBitWidth())
-          buff.insertBits(drive.second, drive.first);
+          curr.size * 8,
+          llvm::makeArrayRef(reinterpret_cast<uint64_t *>(curr.value.get()),
+                             llvm::divideCeil(curr.size, 8)));
+
+      // Apply the changes to the buffer until we reach the next signal.
+      while (i < e && pop.changes[i].first == sigIndex) {
+        const auto &change = pop.buffers[pop.changes[i].second];
+        const auto offset = change.first;
+        const auto &drive = change.second;
+        if (drive.getBitWidth() < buff.getBitWidth())
+          buff.insertBits(drive, offset);
         else
-          buff = drive.second;
+          buff = drive;
+
+        ++i;
       }
 
       // Skip if the updated signal value is equal to the initial value.
-      if (std::memcmp(curr->value.get(), buff.getRawData(), curr->size) == 0)
+      if (std::memcmp(curr.value.get(), buff.getRawData(), curr.size) == 0)
         continue;
 
       // Apply the signal update.
-      std::memcpy(curr->value.get(), buff.getRawData(),
-                  state->signals[change.first].size);
+      std::memcpy(curr.value.get(), buff.getRawData(), curr.size);
 
       // Add sensitive instances.
-      for (auto inst : state->signals[change.first].triggers) {
+      for (auto inst : curr.triggers) {
         // Skip if the process is not currently sensible to the signal.
         if (!state->instances[inst].isEntity) {
-          auto &sensList = state->instances[inst].sensitivityList;
+          const auto &sensList = state->instances[inst].sensitivityList;
           auto it = std::find_if(sensList.begin(), sensList.end(),
-                                 [&change](SignalDetail &s) {
-                                   return s.globalIndex == change.first;
+                                 [sigIndex](const SignalDetail &sig) {
+                                   return sig.globalIndex == sigIndex;
```

Comment:
I wonder if -- in the long run -- we could cook up some sort of sparse bitset/bitvec data structure that has an efficient intersect/and operation, such that we could process all changes, set the bits corresponding to the changed signals, and then intersect that with the sensitivity bits of the instances.

The current implementation looks very reasonable 😃

Reply:
True, that seems like a great idea! It would also split the processing from the tracing/determining what instances have to wakeup, probably making the whole thing more readable and clear 😃